### ISSUE TITLE
TypeError: self.envelopeTemplateDefinition.constructFromObject is not a function

### ISSUE THREAD

**Issue by vasyldemchuk** [docusign/docusign-node-client#52]
I am getting error when use TemplateApi.get method:

TypeError: self.envelopeTemplateDefinition.constructFromObject is not a function

  at EnvelopeTemplate.self.constructFromObject (\node_modules\docusign-esign\src\model\EnvelopeTemplate.js:304:41)
  at handleResponse (\node_modules\docusign-esign\src\api\TemplatesApi.js:183:20)
  at \node_modules\docusign-esign\src\ApiClient.js:215:9
  at Request.callback (\node_modules\docusign-esign\node_modules\superagent\lib\node\index.js:816:12)
  at IncomingMessage.<anonymous> (\node_modules\docusign-esign\node_modules\superagent\lib\node\index.js:1046:12)
  at emitNone (events.js:72:20)
  at IncomingMessage.emit (events.js:166:7)

**Comment by nicholasareed** [comment]
@vasyldemchuk  Can you post a little more of your code that calls the TemplateApi?

**Comment by atheros** [comment]
Example code: http://pastebin.com/kD22fDm0
It tries to get the first template from what was returned by listTemplates()

templateApi.get() throws this error:
```

/xxx/docusign-node-test/node_modules/docusign-esign/src/model/EnvelopeTemplate.js:303
        self.envelopeTemplateDefinition.constructFromObject(data.envelopeTemplateDefinition);
                                        ^

TypeError: self.envelopeTemplateDefinition.constructFromObject is not a function
    at EnvelopeTemplate.self.constructFromObject (/xxx/docusign-node-test/node_modules/docusign-esign/src/model/EnvelopeTemplate.js:303:41)
    at handleResponse (/xxx/docusign-node-test/node_modules/docusign-esign/src/api/TemplatesApi.js:183:20)
    at /xxx/docusign-node-test/node_modules/docusign-esign/src/ApiClient.js:215:9
    at Request.callback (/xxx/docusign-node-test/node_modules/superagent/lib/node/index.js:816:12)
    at IncomingMessage.<anonymous> (/xxx/docusign-node-test/node_modules/superagent/lib/node/index.js:1046:12)
    at emitNone (events.js:91:20)
    at IncomingMessage.emit (events.js:185:7)
    at endReadableNT (_stream_readable.js:975:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)

```

tested on a mac, docusign-esign version is 2.0.3

**Comment by nicholasareed** [comment]
TemplateApi should be creating an `EnvelopeTemplateDefinition`, instead of an `EnvelopeTemplate` inside the `.get`. 

For a temporary fix, you can modify the top of `/docusign-esign/src/api/TemplatesApi.js` to include `EnvelopeTemplateDefinition`: 

```
(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(['../Configuration', '../ApiClient', '../model/EnvelopeTemplateResults', '../model/ErrorDetails', '../model/TemplateSummary', '../model/EnvelopeTemplate', '../model/EnvelopeTemplateDefinition', '../model/TemplateUpdateSummary', '../model/CustomFields', '../model/TemplateCustomFields', '../model/TemplateDocumentsResult', '../model/EnvelopeDefinition', '../model/DocumentFieldsInformation', '../model/PageRequest', '../model/LockInformation', '../model/Notification', '../model/TemplateNotificationRequest', '../model/Recipients', '../model/RecipientsUpdateSummary', '../model/TemplateRecipients', '../model/Tabs', '../model/TemplateTabs', '../model/GroupInformation'], factory);
  } else if (typeof module === 'object' && module.exports) {
    // CommonJS-like environments that support module.exports, like Node.
    module.exports = factory(require('../Configuration'), require('../ApiClient'), require('../model/EnvelopeTemplateResults'), require('../model/ErrorDetails'), require('../model/TemplateSummary'), require('../model/EnvelopeTemplate'), require('../model/EnvelopeTemplateDefinition'), require('../model/TemplateUpdateSummary'), require('../model/CustomFields'), require('../model/TemplateCustomFields'), require('../model/TemplateDocumentsResult'), require('../model/EnvelopeDefinition'), require('../model/DocumentFieldsInformation'), require('../model/PageRequest'), require('../model/LockInformation'), require('../model/Notification'), require('../model/TemplateNotificationRequest'), require('../model/Recipients'), require('../model/RecipientsUpdateSummary'), require('../model/TemplateRecipients'), require('../model/Tabs'), require('../model/TemplateTabs'), require('../model/GroupInformation'));
  } else {
    // Browser globals (root is window)
    if (!root.Docusign) {
      root.Docusign = {};
    }
    root.Docusign.TemplatesApi = factory(root.Docusign.Configuration, root.Docusign.ApiClient, root.Docusign.EnvelopeTemplateResults, root.Docusign.ErrorDetails, root.Docusign.TemplateSummary, root.Docusign.EnvelopeTemplate, root.Docusign.EnvelopeTemplateDefinition, root.Docusign.TemplateUpdateSummary, root.Docusign.CustomFields, root.Docusign.TemplateCustomFields, root.Docusign.TemplateDocumentsResult, root.Docusign.EnvelopeDefinition, root.Docusign.DocumentFieldsInformation, root.Docusign.PageRequest, root.Docusign.LockInformation, root.Docusign.Notification, root.Docusign.TemplateNotificationRequest, root.Docusign.Recipients, root.Docusign.RecipientsUpdateSummary, root.Docusign.TemplateRecipients, root.Docusign.Tabs, root.Docusign.TemplateTabs, root.Docusign.GroupInformation);
  }
}(this, function(Configuration, ApiClient, EnvelopeTemplateResults, ErrorDetails, TemplateSummary, EnvelopeTemplate, EnvelopeTemplateDefinition, TemplateUpdateSummary, CustomFields, TemplateCustomFields, TemplateDocumentsResult, EnvelopeDefinition, DocumentFieldsInformation, PageRequest, LockInformation, Notification, TemplateNotificationRequest, Recipients, RecipientsUpdateSummary, TemplateRecipients, Tabs, TemplateTabs, GroupInformation) {
  'use strict';

  var TemplatesApi = function TemplatesApi(apiClient) {
    this.apiClient = apiClient || Configuration.default.getDefaultApiClient();

    var self = this;
    
    .....(hid the rest)
    
```

and further down the file (~ line 183), change 

`var result = new EnvelopeTemplate();` 

to 

`var result = new EnvelopeTemplateDefinition();`


I was able to get the Template after doing this.

**Comment by atheros** [comment]
It still doesn't work.

I'm getting a nice JSON object from server (http://pastebin.com/eZrxJ0K9), but the response from the SDK is empty:
```
EnvelopeTemplateDefinition {
  templateId: null,
  name: null,
  shared: null,
  password: null,
  newPassword: null,
  description: null,
  lastModified: null,
  lastModifiedBy: null,
  pageCount: null,
  uri: null,
  folderName: null,
  folderId: null,
  folderUri: null,
  parentFolderUri: null,
  owner: null,
// followed by lots of functions
```

**Comment by ejhayes** [comment]
I'm running into this same issue as well.  I'm not doing anything fancy to call this either:

```
templatesApi.get(accountId, templateId, function(error, data, response) {
  // fails before getting to this point
});
```

I checked the api explorer and it appears to be returning a Envelope TemplateDefinition: https://apiexplorer.docusign.com/#/?categories=Templates&tags=Templates&operations=get

**Comment by atheros** [comment]
As far as I understand, this is the OFFICIAL SDK from DokuSign, right? So do the authors understand their SDK is currently broken and not working?

**Comment by mmallis87** [comment]
The workaround for now is to use the API explorer to generate the raw API call and parse the response json yourself, until this bug is fixed.

**Comment by szb0018** [comment]
The following code snippet is the culprit:
`if (data.envelopeTemplateDefinition) {
        self.envelopeTemplateDefinition = new data.envelopeTemplateDefinition.constructor();
        self.envelopeTemplateDefinition.constructFromObject(data.envelopeTemplateDefinition);
      }`

In the Api result handler, a new model instance is created and then populated by calling the constructFromObject method and providing the POJO returned from the REST API. The issue occurs in this function when a sub-model is encountered because the function assumes that the data type is a model when in fact it is just another POJO.

This assumption does not make much sense though, since the whole point of the function is to populate the model from an Object - not another model. Converting each occurrence to the appropriate model instantiation resolved the issue and returns a populated model.

I would love to hear some input from the developers on why this approach was taken, since it is used throughout the entire code base and appears to be deliberate and not an oversight. Is there a another handler that is supposed to pre-process the responses and swap out the constructor methods recursively?

For now, I am creating a fork and will be submitting a pull request to fix this in every model. I'll provide an update with a link to the fork when it's done.

**Comment by szb0018** [comment]
This fork contains the fix mentioned above. Hopefully it will get merged into master here soon.
szb0018/docusign-node-client

**Comment by mmallis87** [comment]
Fixed in 3.0.0

**Comment by alexesca** [comment]
Need clarification. Code example here in the API docs (https://bit.ly/33LyZr0) should be updated or this model(EnvelopeTemplateDefinition) should be made available.  The code example in the api docs says to use this model, but the code luncher creates templates without the `EnvelopeTemplateDefinition`. Not sure if a new issue should be created or addressed here. I installed docusign-esign 5.3.0 and 5.6.0 and `const envelopeTemplateDefinition = new docusign.EnvelopeTemplateDefinition.constructFromObject(...` fails. The model is found int the repo, but not accessible from `const docusign = require('docusign-esign') ... docusign.EnvelopeTemplateDefinition({..})`
